patches/chromium/PipeWire-capturer-implement-proper-DMA-BUFs-support.patch

```diff
@@ -0,0 +1,935 @@
++    if (*it == "EGL_EXT_platform_base") {
++      has_platform_base_ext = true;
++      continue;
++    } else if (*it == "EGL_MESA_platform_gbm") {
++      has_platform_gbm_ext = true;
++      continue;
++    }
++  }
++
++  if (!has_platform_base_ext || !has_platform_gbm_ext) {
++    RTC_LOG(LS_ERROR) << "One of required EGL extensions is missing";
++    return;
++  }
++
++  // Use eglGetPlatformDisplayEXT() to get the display pointer
++  // if the implementation supports it.
++  egl_.display =
++      eglGetPlatformDisplayEXT(EGL_PLATFORM_GBM_MESA, gbm_device_, nullptr);
++
++  if (egl_.display == EGL_NO_DISPLAY) {
++    RTC_LOG(LS_ERROR) << "Error during obtaining EGL display: "
++                      << FormatGLError(eglGetError());
++    return;
++  }
++
++  EGLint major, minor;
++  if (eglInitialize(egl_.display, &major, &minor) == EGL_FALSE) {
++    RTC_LOG(LS_ERROR) << "Error during eglInitialize: "
++                      << FormatGLError(eglGetError());
++    return;
++  }
++
++  if (eglBindAPI(EGL_OPENGL_API) == EGL_FALSE) {
++    RTC_LOG(LS_ERROR) << "bind OpenGL API failed";
++    return;
++  }
++
++  egl_.context =
++      eglCreateContext(egl_.display, nullptr, EGL_NO_CONTEXT, nullptr);
++
++  if (egl_.context == EGL_NO_CONTEXT) {
++    RTC_LOG(LS_ERROR) << "Couldn't create EGL context: "
++                      << FormatGLError(eglGetError());
++    return;
++  }
++
++  const char* client_extensions_cstring_display =
++      eglQueryString(egl_.display, EGL_EXTENSIONS);
++  client_extensions_string = client_extensions_cstring_display;
++
++  bool has_image_dma_buf_import_ext = false;
++  bool has_image_dma_buf_import_modifiers_ext = false;
++
++  for (auto it = egl_.extensions.begin(); it != egl_.extensions.end(); it++) {
++    if (*it == "EGL_EXT_image_dma_buf_import") {
++      has_image_dma_buf_import_ext = true;
++      continue;
++    } else if (*it == "EGL_EXT_image_dma_buf_import_modifiers") {
++      has_image_dma_buf_import_modifiers_ext = true;
++      continue;
++    }
++  }
++
++  if (has_image_dma_buf_import_ext && has_image_dma_buf_import_modifiers_ext) {
++    EglQueryDmaBufFormatsEXT = (eglQueryDmaBufFormatsEXT_func)eglGetProcAddress(
++        "eglQueryDmaBufFormatsEXT");
++    EglQueryDmaBufModifiersEXT =
++        (eglQueryDmaBufModifiersEXT_func)eglGetProcAddress(
++            "eglQueryDmaBufModifiersEXT");
++  }
++
++  RTC_LOG(LS_INFO) << "Egl initialization succeeded";
++  egl_initialized_ = true;
++}
++
++std::unique_ptr<uint8_t[]> EglDmaBuf::ImageFromDmaBuf(int32_t fd,
++                                                      const DesktopSize& size,
++                                                      int32_t stride,
++                                                      uint32_t format,
++                                                      uint32_t offset,
++                                                      uint64_t modifier) {
```

Comment:
Please extend fd, stride and offset to an array, so we can import buffers with multiple planes.

Reply:
Should be now implemented, can you check? One thing I'm not sure about is that later I initialize `src` with:

`src = std::make_unique<uint8_t[]>(strides[0] * size.height());`

Will the stride be same for each plane or should I just get the maximum stride from all planes and use that one?